torch/csrc/Module.cpp

```diff
@@ -524,6 +529,58 @@ PyObject *THPModule_isEnabledXNNPACK(PyObject * /* unused */)
   else Py_RETURN_FALSE;
 }
 
+namespace {
+struct WatchdogThread {
+  WatchdogThread(const std::string& name_, int timeout): name(name_), thrd([this, timeout] {
+      using namespace std::chrono_literals;
+      std::unique_lock<std::mutex> lock(mutex);
+      cv.wait_for(lock, timeout *1s, [this] { return exitFlag; });
+      if (!exitFlag) {
+        std::cerr << "Test " << name << " takes longer than " << timeout << " seconds" <<std::endl;
+        std::exit(-1);
+      }
+      }) {}
+  ~WatchdogThread() {
+    {
+      std::lock_guard<std::mutex> guard(mutex);
+      exitFlag = true;
+    }
+    cv.notify_one();
+    if (thrd.joinable())
+      thrd.join();
+  }
+private:
+  std::string name;
+  std::mutex mutex;
+  std::condition_variable cv;
+  bool exitFlag = false;
+  // Thread must be the last member, otherwise it will not get properly initialized
+  std::thread thrd;
+};
+}
+
+PyObject *THPModule_setWatchdog(PyObject *_unused, PyObject *args)
+{
+  static std::unordered_map<std::string, std::unique_ptr<WatchdogThread>> watchdogs;
+  char *cname = nullptr;
+  unsigned timeout = 0;
+  if (!PyArg_ParseTuple(args, "sI", &cname, &timeout)) {
+    return nullptr;
+  }
+  std::string name(cname);
+  auto it = watchdogs.find(name);
+  if (it != watchdogs.end()) {
+   if (timeout == 0) {
+     watchdogs.erase(it);
+     Py_RETURN_NONE;
+   }
+   std::cerr << " Ignoring watchdog override " << name << std::endl;
```

Comment:
lol

Reply:
Can we just make this a hard error?

Reply:
Don't have strong preference, one way or the other, will add `std::abort()` and see what will happen.